[EuiDataGrid] Fix row height tests and mocks

weltenwort opened the pull request:
## :memo: Summary

This refactors the `RowHeightUtils` tests to fix problems stemming from spies and fake timers. This has been split out of #6028.

## :female_detective: Review notes

- I refactored the way the `RowHeightUtils` receive references to various dependencies. It now receives them as `MutableRefObject`s in its constructor without the need for several imperative `set*` methods. This will make it easier to extend and mock in the future.
- The tests for the `RowHeightUtils` were not working as intended, because they used some jest features incorrectly. I tried to correct for the following problems before adding additional test cases for the new `compensateForLayoutShift` method:
  - `jest.useFakeTimers` was called in several places directly in `describe` blocks, which means they're not bound to the lifecycle of the test cases. As a consequence some tests failed unpredictably depending on the execution order of the `describe` blocks. Some test cases also were not calling `jest.runAllTimers`, which caused several `setTimeouts` not to be executed and lead to several false positive test results.
  - `jest.spyOn` was also called in `describe` blocks directly, but their implementation was never restored. That means that some spied-upon function implementations would be replaced by no-ops when the mocks were reset, which in turn lead to false positives and negatives.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6050/

constancecchen commented on src/components/datagrid/utils/row_heights.ts:
```diff
@@ -259,19 +258,17 @@ export const useRowHeightUtils = ({
-  useEffect(() => {
-    if (gridRef) rowHeightUtils.setGrid(gridRef);
-    rowHeightUtils.setRerenderGridBody(forceRender);
-  }, [gridRef, forceRender, rowHeightUtils]);
+  const forceRenderRef = useLatest(forceRender);
+  const rowHeightUtils = useMemo(
+    () => new RowHeightUtils(gridRef, forceRenderRef),
+    [forceRenderRef, gridRef]
+  );
```
++ to passing args to `new RowHeightUtils` rather than using `setX` methods - that's a super nice cleanup, thank you!

I'm not super following `useLatest()` - it looks like all it's doing is basically turning `forceRender` into a ref. Is it not possible to simply do something like 

```jsx
const forceRenderRef = useRef(useForceRender());
```

?

Or alternatively, why pass `forceRender` as a ref at all? Can we not simply pass `forceRender` as-is without a ref?

constancecchen replied:
Gotcha, that makes sense.

I think my main question here is the addition of the `useLatest` hook/file - it's still not clear to me what that util does that `useRef` doesn't. If we had to separate `forceRender` and `forceRenderRef`, does

```js
const forceRender = useForceRender();
const forceRenderRef = useRef(forceRender);
```

not work instead?

> The reason for assigning to forceRender instead of doing it in one line is that is that the code further down uses forceRender directly as a useEffect dependency.

FWIW, we can easily change this useEffect dependency to use the ref instead of `forceRender` directly. That useEffect doesn't actually need forceRender as a "change when this changes" dependency, it just happens to rely on using the fn, so the linter wants it in the dependency array.

Between the two code snippets I mentioned I think I'd rather just use `useRef(useForceRender());` and update all downstream components to call the `ref.current`.

> The reason to pass it to the RowHeightUtils as a ref is that we don't want to re-instantiate RowHeightUtils just because forceRender changes identity

For what it's worth also, `forceRender` never changes identity. It's wrapped in a useCallback with an empty dependency array:

https://github.com/elastic/eui/blob/b97f93e392694bb1b0baf3bba7da2a81321eaba0/src/services/hooks/useForceRender.ts#L13-L15

So inasmuch as we can rely on `useCallback` to remain referentially intact, (which to be fair `useRef` is more reliable for, so I'm totally ok using it as an extra layer of insurance), we can be fairly confident `forceRender` will not cause a re-instantiation.

weltenwort replied:
Switching to using the `forceRender` ref everywhere makes sense. :+1: The difference between `const forceRenderRef = useRef(forceRender)` and `const forceRenderRef = useLatest(forceRender)` is, that the former only sets `forceRenderRef.current` once initially, while the latter updates it with the latest `forceRender` every time.

That neatly transitions to the question of the stability of `useCallback`. It just being syntactic sugar for `useMemo` and the React docs stating "You may rely on useMemo as a performance optimization, not as a semantic guarantee." [[1]](https://reactjs.org/docs/hooks-reference.html#usememo) implies we shouldn't rely on it. But to be consistent with that, we also shouldn't rely on `useMemo` in

```typescript
  const rowHeightUtils = useMemo(
    () => new RowHeightUtils(gridRef, forceRenderRef),
    [forceRenderRef, gridRef]
  );
```

Instead we should probably do something like

```typescript
  const [rowHeightUtils] = useState(() => new RowHeightUtils(gridRef, forceRenderRef));
```

to ensure it's only instantiated once.

chandlerprall replied:
I've missed having Felix around 😀 

The `useLatest` ref path solves for the reason we opted for the useEffect to set `gridRef` and `forceRender` fns, so I'm very happy to see avoiding a re-instantiation of `RowHeightUtils` was already considered and I didn't have to remember all the reasons why ❤️ 

> If we do so, can we actually remove the useLatest/ref around forceRender in that case?

Still need to update the gridRef/forceRender values when they change so the once-instantiated RowHeightUtils is kept current (`current`, 😆).

+1 for the useState approach for the reasons given. Almost certain we have made the same decision elsewhere when we _cannot_ be okay with useMemo forgetting something.

constancecchen replied:
> Still need to update the gridRef/forceRender values when they change so the once-instantiated RowHeightUtils is kept current (current, 😆).

gridRef, sure. forceRender as I mentioned is a memoized set state fn and should never really change or need to be kept current. You can see this in play when you do something like:

```jsx
const [foo, setFoo] = useState();
const useEffect(() => {
  // do something
}, []); // setFoo isn't required as a dependency by react's linters because it never updates
```

Based on that, it's not clear to me why we would need to store it as a ref.

constancecchen replied:
Ahh gotcha, you're totally right on that, apologies. Would `useRef` not still work on that? Is there a specific need for a new `useLatest` util? If the answer is yes, can I ask we do 2 things in the `useLatest` file:

1. Write unit tests for `useLatest`, 
2. Make it clear via either tests or a code block comment what the use case / difference is between `useRef` and `useLatest` (I'll be honest, if I was just passing through that distinction wouldn't be clear to me)

constancecchen replied:
Thanks a million!!

constancecchen commented on src/components/datagrid/utils/row_heights.ts:
```diff
@@ -36,7 +38,14 @@ export const cellPaddingsMap: Record<EuiDataGridStyleCellPaddings, number> = {
 export const AUTO_HEIGHT = 'auto';
 export const DEFAULT_ROW_HEIGHT = 34;
 
+type IGrid = Omit<Grid, keyof Component>;
```
I'm not quite following why this` IGrid` type is needed - can you clarify?

weltenwort replied:
`Grid` is a React class component with all the private methods like `setState` and so on. But we're only interested in the public, imperative API, but `react-window` exports no separate type for that. Using the `Omit` type we can derive that public interface, though. This makes mocking without a full component a lot easier.

constancecchen replied:
TIL! Thank you for explaining super clearly. I think it might be worth noting this in an inline code comment in the event that react-window someday exports a separate type - up to you though

constancecchen commented on src/components/datagrid/utils/__mocks__/row_heights.ts:
```diff
@@ -8,26 +8,38 @@
-  isRowHeightOverride: jest.fn(actual.isRowHeightOverride),
-  setRerenderGridBody: jest.fn(),
-} as unknown) as ActualRowHeightUtils;
+export const RowHeightUtils = jest
+  .fn<
+    ActualRowHeightUtils,
+    ConstructorParameters<typeof ActualRowHeightUtils>
+  >()
+  .mockImplementation((...args) => {
+    const rowHeightUtils = new ActualRowHeightUtils(...args);
```
Super cool, and I definitely see how this fixes issues of stale instances between tests. Thanks for making this change!

constancecchen commented on src/components/datagrid/utils/row_heights.test.ts:
```diff
@@ -374,54 +405,73 @@ describe('RowHeightUtils', () => {
+    current: {
+      resetAfterIndices: jest.fn(),
+      resetAfterColumnIndex: jest.fn(),
+
```
nit: unintentional newline?
```suggestion
```

weltenwort replied:
indeed, good catch

constancecchen commented:
@weltenwort After more thorougly reviewing the diff, I'd vote no changelog needed - the small amount of source code being changed could be classified as tech debt only / non-consumer-impacting.

I think I've left all the comments I'm going to, though would def appreciate a second eye from @chandlerprall. My main/only semi-blocking question would be around `useLatest` - I'm not quite fully seeing/understanding the need for it. Everything else is a minor non blocking comment or question. Thanks again, this is great!! 👏

weltenwort commented on src/components/datagrid/utils/row_heights.ts:
```diff
@@ -36,7 +38,14 @@ export const cellPaddingsMap: Record<EuiDataGridStyleCellPaddings, number> = {
+type IGrid = Omit<Grid, keyof Component>;
+
 export class RowHeightUtils {
+  constructor(
+    private gridRef: MutableRefObject<IGrid | null>,
+    private rerenderGridBodyRef: MutableRefObject<(() => void) | null>
+  ) {}
```
yes, that's exactly what happens :+1:

weltenwort commented on src/components/datagrid/utils/row_heights.test.ts:
```diff
@@ -6,18 +6,38 @@
+
+  afterEach(() => {
+    jest.useRealTimers();
+    jest.resetAllMocks();
```
I should work, but I'll have to test whether it breaks anything. Doing the individual `mockRestore()` near each spy creation could be preferable for readability, though. The global restoration is easy to lose track of.

constancecchen replied:
That makes sense as a reason - if you prefer the explicitness I don't have a strong preference either way, so would be good with leaving it in as-is

weltenwort commented:
@constancecchen thank you for the thorough review. I tried to respond to your comments inline. Let me know if those explanations make sense.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6050/

constancecchen commented on src/components/datagrid/body/data_grid_body.test.tsx:
```diff
@@ -9,12 +9,22 @@
+  };
+  const gridItemsRendered = { current: null };
+  const rerenderGridBodyRef = { current: null };
+  const rowHeightUtils = new RowHeightUtils(gridRef, rerenderGridBodyRef);
```
Super interesting!! (I love code discussions like this BTW, so you'll have to forgive me for overindulging 😄)

> allows for more compiler checks and JIT optimizations.

TIL 🎓 I had thought that instantiating fewer consts/vars vs. just 1 obj would be 'better' in terms of optimization. Compiler checks makes sense though. I def know we're splitting hairs at this point though in terms of microperf, but definitely food for me to think about!

> It usually reduces cognitive load on the reader

I definitely acknowledge this is very subjective, but my personal 2c is that more cruft/setup at the top of the file is harder for me to read / more cognitive load than otherwise. I prefer to jump straight into the core purpose/logic of the file when possible to try and grasp its intentions. When I come across a var I'll tend to store it in my short-term memory assuming it'll be re-used later and I'll want to try and connect the var definition/reference to where it's being used, so more vars = more things flying around my head to try and remember. So for the test files if I start by seeing a single `const requiredProps = ...` declaration, I get a sense of "oh these are just the props needed to get the test started", and I'll collapse that obj or skim/skip past it, whereas if I see other vars defined I'll be like "this seems important/will come into play later, I'll try to remember it individually".

Like I said, very much subjective though, forgive my rambling! My comment definitely was optional/not a change request so if you prefer it this way, feel free to leave it

weltenwort commented:
@constancecchen, @chandlerprall thanks for all the feedback! I have changed the `RowHeightUtils` instantiation to happen inside of a `useState` initializer, removed the bare usage of `forceRender` and added tests for `useLatest`. Let me know what you think.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6050/

constancecchen commented on src/services/hooks/useLatest.ts:
```diff
@@ -0,0 +1,19 @@
+
+/**
+ * Wraps the given `value` into a `MutableRefObject` and keeps the `current`
+ * value up-to-date on very render cycle.
```
Super minor copy nit, also optional - feel free to merge w/o
```suggestion
 * value up-to-date on every render cycle.
```

constancecchen reviewed (approved):
Fantastic! Thank you so much for taking the time to explain how all these changes worked, it really helped a lot!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_6050/
